### ISSUE TITLE
Dark Heresy 1st Edition

### ISSUE THREAD

**Issue by RobbaYaga** [Roll20/roll20-character-sheets#7301]
There are a couple problems that have appeared since the character sheets were recently updated:

1) The dice roller now gives decimals instead of whole numbers when indicating levels of success. For instance, a roll that might generate 1 level of success now generates 1.2 levels of success.

2) The new range categories on the ranged weapons are incorrect. These range categories never existed on the official character sheet, but are a part of the game mechanics and are as follows:

Point Blank (PB) - usually 3m or less
Short (S) - less than half the weapon's listed range
Medium (M) - Anything from half the weapon's listed range up to twice the weapon's listed range
Long (L) - greater than twice the weapon's listed range
Extreme (E) - greater than three times the weapon's listed range.

Thank you,
RY

**Comment by Stexinator** [comment]
Is this still an issue? I am currently working on 2ed and could take a look if it is a quick fix. Can you please create an issue for each point?

**Comment by Stexinator** [comment]
I dont understand the second point. Looking at the character sheet the placeholder gives the values @RobbaYaga provided

**Comment by RobbaYaga** [comment]
Yeah both are still an issue:

1) This is apparently a problem with it being difficult to program the system to to round down when one gets a success (2.2 = round down to 2 levels of sux) and round up when one fails (2.2 rounds up to 2 levels of failure).

2) This remains unchanged and incorrect

3) This is new but at various times, the system logs something as a success or failure when the dice roll shown indicates the opposite. I don't have a specific example, but it appears most often that certain successes show up as rather dramatic failures. Not sure if this is related to the rounding issues.

Thanks for looking at it.

**Comment by Stexinator** [comment]
Since I am not familiar with the first edition ruling I need some more clarification. In 2ed it is just the tenth digit that is regarded, when I looked at the roll-template the order of rounding was wrong, so this might be the issue here.

@RobbaYaga Correct me if I am wrong: 

- Roll 10 Target 22 => 1 Degrees of Success
- Roll 20 Target 22 => 0 Degrees of Success
- Roll 10 Target 29 => 1 Degrees of Success
- Roll 19 Target 20 => 1 Degrees of Success
- Roll 19 Target 29 => 1 Degrees of Success
- Roll 19 Target 19 => 0 Degrees of Success
- Roll 10 Target 19 => 0 Degrees of Success

- Roll 20 Target 10 => 1 Degrees of Failure
- Roll 20 Target 19 => 1 Degrees of Failure
- Roll 29 Target 10 => 1 Degrees of Failure

**Comment by Stexinator** [comment]
@RobbaYaga can you extract the range issue into a separate ticket and edit your starting post?

**Comment by Stexinator** [comment]
This links to https://github.com/Roll20/roll20-character-sheets/pull/7195

**Comment by RobbaYaga** [comment]
I have created two new tickets, one for the Range Issue and one for the Incorrect or Missing Dice Results issue.

**Comment by RobbaYaga** [comment]
> Since I am not familiar with the first edition ruling I need some more clarification. In 2ed it is just the tenth digit that is regarded, when I looked at the roll-template the order of rounding was wrong for 2nd, so this might be the issue here.
> 
> @RobbaYaga Correct me if I am wrong:
> 
> * Roll 10 Target 22 => 1 Degrees of Success (YES)
> * Roll 20 Target 22 => 0 Degrees of Success (YES)
> * Roll 10 Target 29 => 1 Degrees of Success (YES)
> * Roll 19 Target 20 => 1 Degrees of Success (No, this would be 0 Degrees of Success like the second example above)
> * Roll 19 Target 29 => 1 Degrees of Success (Yes)
> * Roll 19 Target 19 => 0 Degrees of Success (No, this would be 1 Degrees of Success)
> * Roll 10 Target 19 => 0 Degrees of Success (YES)
> * Roll 20 Target 10 => 1 Degrees of Failure (YES)
> * Roll 20 Target 19 => 1 Degrees of Failure (No, this would be 0 Degrees of Failure)
> * Roll 29 Target 10 => 1 Degrees of Failure (YES)

In DH1e, "Degrees of Success" or "Degrees of Failure" really refers to Degrees of EXTRA Success or EXTRA Failure. If the roll is with 9 of the target, it is simply a success or failure. As soon as you have a difference of 10, it becomes an (extra) level of success or failure.

Difference of 0-9 = 0 levels of Success / Failure
Difference of 10-19 = 1 level of Success / Failure
Difference of 20-29 = 2 levels of Success / Failure
etc.

As stated before, the rounding issue is a problem because:
Successes are rounded down: Target of 50, roll of 38 = 1.2 levels of success which is rounded down to 1 level of success
Failures are rounded up: Target of 50, roll of 62 = 1.2 levels of failure which is rounded up to 1 level of failure.

The problem seems to be the negative values associated with calculating the difference between roll and target. If there is a way to take that difference, determine if it is a success or failure and then use an absolute value before rounding, it should solve the issue. I don't know JS however, so not sure how involved that is.

**Comment by Stexinator** [comment]
@nmbradley This isssue cannot be resolved until Roll20 changes things in the dice engine

**Comment by nmbradley** [comment]
Good to know! Can I get some more clarity on that response?

Is this because of a rounding issue? Or because of the way that dice are handled?

Either way, I'll move this to an internal ticket, but any information you can provide would be useful.

**Comment by Stexinator** [comment]
The logic used in the sheet works with saved dice values. The issue is that it can be rounded only once, when making the actual roll, due to not being able to calculate with saved values.

As soon as saved dice rolls can be rounded this issue can be resolved. Other sheets have similar issues with multiplication, so this would be a feature that can be used for a broader audience

**Comment by nmbradley** [comment]
Perfect, thank you. 

I'm going to close this ticket as it only pertains to an issue with the VTT rather than the sheet itself. Feel free to create a new issue if you think that there are residual issues that still need to be addressed on the sheet side of things.
